### ISSUE TITLE
DataSource metrics are not available when DataSourceAutoConfiguration is excluded and a user-defined DataSource is available

### ISSUE THREAD

**Issue by 730alchemy** [spring-projects/spring-boot#13295]
Spring Boot 2.0.2.RELEASE

I have a Spring Boot application configured to use JPA and to send metrics to New Relic.  New Relic is receiving metrics from the application, but not Hikari connection metrics.  

I used Postman to send a request to the application "metrics" endpoint.  The list of metrics returned did not include any of the "hikari.connections" metrics.  The list also did not include any "jdbc.connections" metrics.  The list does include other metrics, such as "jvm." and "rabbitmq.", so the endpoint is working and micrometer appears to be providing some metrics (and like I said above, New Relic confirms this).

When starting the application, the console indicates that HikariCP-2.7.9.jar is on the classpath.  The application is able to successfully connect to a database and the repositories work correctly.

A trace on OnClassCondition shows this line in the console:
```
8:39:04.807 TRACE 326836 --- [  restartedMain] o.s.b.a.condition.OnClassCondition       : Condition OnClassCondition on org.springframework.boot.actuate.autoconfigure.metrics.jdbc.DataSourcePoolMetricsAutoConfiguration$HikariDataSourceMetricsConfiguration matched due to @ConditionalOnClass found required class 'com.zaxxer.hikari.HikariDataSource'; @ConditionalOnMissingClass did not find unwanted class
```
The console also indicates that a HikariDataSource is successfully started
```
2018-05-29 18:39:08.801  INFO 326836 --- [ost-startStop-1] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Starting...
2018-05-29 18:39:08.967  INFO 326836 --- [ost-startStop-1] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Start completed.
```

The remaining Hikari outputs in the log are these:
```
2018-05-29 18:39:14.320  INFO 326836 --- [  restartedMain] o.s.j.e.a.AnnotationMBeanExporter        : Bean with name 'hikariConfig' has been autodetected for JMX exposure
2018-05-29 18:39:14.320  INFO 326836 --- [  restartedMain] o.s.j.e.a.AnnotationMBeanExporter        : Bean with name 'dataSource' has been autodetected for JMX exposure
2018-05-29 18:39:14.327  INFO 326836 --- [  restartedMain] o.s.j.e.a.AnnotationMBeanExporter        : Bean with name 'rabbitConnectionFactory' has been autodetected for JMX exposure
2018-05-29 18:39:14.327  INFO 326836 --- [  restartedMain] o.s.aop.framework.CglibAopProxy          : Method [void com.zaxxer.hikari.HikariConfig.seal()] is package-visible across different ClassLoaders and cannot get proxied via CGLIB: Declare this method as public or protected if you need to support invocations through the proxy.
2018-05-29 18:39:14.715  INFO 326836 --- [  restartedMain] o.s.j.e.a.AnnotationMBeanExporter        : Located MBean 'dataSource': registering with JMX server as MBean [com.zaxxer.hikari:name=dataSource,type=HikariDataSource]
```

Let me know what other information you need. Thanks.

**Comment by nosan** [comment]
Could you please try to use a version 3.0.0?

**Comment by wilkinsona** [comment]
@nosan Thanks for the suggestion but that fix for that issue is in Hikari 2.7.9 so there should be no need to try 3.0.

@730alchemy I can't reproduce the problem based on the information you've provided. A minimal application using Spring Boot 2.0.2 lists the following metrics:

```
http :8080/actuator/metrics
HTTP/1.1 200
Content-Type: application/vnd.spring-boot.actuator.v2+json;charset=UTF-8
Date: Wed, 30 May 2018 08:38:26 GMT
Transfer-Encoding: chunked

{
    "names": [
        "jvm.memory.max",
        "http.server.requests",
        "process.files.max",
        "jvm.gc.memory.promoted",
        "tomcat.cache.hit",
        "system.load.average.1m",
        "tomcat.cache.access",
        "jvm.memory.used",
        "jvm.gc.max.data.size",
        "jdbc.connections.max",
        "jdbc.connections.min",
        "jvm.gc.pause",
        "jvm.memory.committed",
        "system.cpu.count",
        "logback.events",
        "tomcat.global.sent",
        "jvm.buffer.memory.used",
        "tomcat.sessions.created",
        "jvm.threads.daemon",
        "system.cpu.usage",
        "jvm.gc.memory.allocated",
        "tomcat.global.request.max",
        "hikaricp.connections.idle",
        "hikaricp.connections.pending",
        "tomcat.global.request",
        "tomcat.sessions.expired",
        "hikaricp.connections",
        "jvm.threads.live",
        "jvm.threads.peak",
        "tomcat.global.received",
        "hikaricp.connections.active",
        "hikaricp.connections.creation",
        "process.uptime",
        "tomcat.sessions.rejected",
        "process.cpu.usage",
        "tomcat.threads.config.max",
        "jvm.classes.loaded",
        "hikaricp.connections.max",
        "hikaricp.connections.min",
        "jvm.classes.unloaded",
        "tomcat.global.error",
        "tomcat.sessions.active.current",
        "tomcat.sessions.alive.max",
        "jvm.gc.live.data.size",
        "tomcat.servlet.request.max",
        "hikaricp.connections.usage",
        "tomcat.threads.current",
        "tomcat.servlet.request",
        "hikaricp.connections.timeout",
        "process.files.open",
        "jvm.buffer.count",
        "jvm.buffer.total.capacity",
        "tomcat.sessions.active.max",
        "hikaricp.connections.acquire",
        "tomcat.threads.busy",
        "process.start.time",
        "tomcat.servlet.error"
    ]
}
```

As you can see there are several `hikaricp.*` and `jdbc.*` metrics listed so there must be something in your setup that you haven't described that is causing the problem. I can see from the log output that you are using DevTools and that something has used the connection pool during startup but attempting to recreate this arrangement hasn't reproduced the problem.

I can also see from the log output that something has triggered proxying of the Hikari DataSource:

```
2018-05-29 18:39:14.327  INFO 326836 --- [  restartedMain] o.s.aop.framework.CglibAopProxy          : Method [void com.zaxxer.hikari.HikariConfig.seal()] is package-visible across different ClassLoaders and cannot get proxied via CGLIB: Declare this method as public or protected if you need to support invocations through the proxy.
```

I can't tell why that might be from what you have described thus far but it could be related to the problem.

Can you please provide a minimal sample that reproduces the problem? As things stand there are too many unknowns for us to make efficient progress on this.

**Comment by 730alchemy** [comment]
I stripped down our project and attached the zip. The parent module contains two modules.  The "master-service" implements the application, so its target folder contains the executable jar. Please run it using Java 8.  Later versions of Java complain that JAXBException class cannot be found.

The application defines two profiles: local and DynamicDB.  The issue occurs under both profiles.  You can run the app using the local profile.  However, you will not be able to run the application using the DynamicDB profile because I had to mangle that code to get a working example to you.

Also, this app reconfigures the management endpoint base path to "/", so the metrics are visible at "/metrics".

[issue-13295.zip](https://github.com/spring-projects/spring-boot/files/2055222/issue-13295.zip)

**Comment by wilkinsona** [comment]
Thanks for the sample. It's revealed rather a lot about your application that wasn't included in your description. The key thing is that you have excluded `DataSourceAutoConfiguration`. This has the side-effect of also disabling the auto-configuration of the beans that provide DataSource-related metrics. We'll fix that by separating that out into its own auto-configuration. In the meantime I believe you can work around the problem you're seeing by removing your exclusion of `DataSourceAutoConfiguration` and instead just rely on the presence of your own `DataSource` bean definition to cause it to back off.

**Comment by wilkinsona** [comment]
Having considered this a bit more, I'm not so sure that we should consider it to be a bug. With the current arrangement there's a difference in behaviour between declaring your own `DataSource` bean and disabling `DataSourceAutoConfiguration` entirely. If we separate out the configuration of the data source meta data providers into a separate auto-configuration that difference will be reduced. There may be users deliberately disabling `DataSourceAutoConfiguration` entirely as they want to make use of that difference. Removing or reducing it would be a breaking change for them.

@730alchemy what was your reason for disabling the auto-configuration? Looking at the code you have shared, removing the exclusion will give you the metrics that you want while also allowing your own data source definition to be used.

**Comment by 730alchemy** [comment]
I removed the exclusion of `DataSourceAutoConfiguration`.  When I call the `metrics` endpoint, I now see the hikari and jdbc connection metrics listed.  However, the application is failing to send the metrics to NewRelic.  When the application attempts to send the metrics to NewRelic, it logs the following error:
``` 
ERROR 448048 --- [pool-1-thread-1] i.m.newrelic.NewRelicMeterRegistry       : failed to send metrics: {"error":"Error parsing JSON payload at event index 8"}
```
The Spring Boot Actuator reference document states that each hikari-specific metric is tagged by the name of the pool, which I was not explicitly configuring.  I thought this might cause the parsing error, so I configured a pool name for the data source, but this did not solve the error ... although it did change the error message from `Error parsing JSON payload at event index 30` to `Error parsing JSON payload at event index 8` ... maybe this fixed one parsing error but another remained?

RE: why we disabled DataSource auto-configuration ... I do not know why this was excluded.  I talked with the original developer, who coded this several months ago, and he can't remember the exact reasons. He remembers the data source auto-config doing something more than creating the data source bean and thus interfering with our custom jpa configuration, but he can't remember the exact problem.

However, that was an older version of Spring Boot, things have changed, improved, etc ... we can remove this exclusion now and re-test to ensure no regression, but we have the JSON parsing issue when transmitting to  NewRelic.  Note that the application does not throw this error when sending a payload to NewRelic that does not contain jdbc and hikari metrics.

**Comment by 730alchemy** [comment]
RE: whether this is a bug ... `DataSource` configuration is orthogonal to `MetricRegistry` configuration.  I do not think that a `DataSource` configuration should interfere with the configuration conventions used for `MetricRegistry`  The current `MetricRegistry` configuration convention, according to the Actuator reference documentation, is to add a registry to the composite `MetricRegistry` for each of the supported implementations that it finds on the classpath.

When `DataSource` auto-configuration is disabled and the application configures its own `DataSource`, the `DataSource` implementation is on the classpath.  Hence, I think the `MetricRegistry` auto-configuration should follow the documented convention and add the appropriate registry.

**Comment by wilkinsona** [comment]
> I do not think that a DataSource configuration should interfere with the configuration conventions used for MetricRegistry.

I agree and, as far as I know, there is no such interference.

> When DataSource auto-configuration is disabled and the application configures its own DataSource, the DataSource implementation is on the classpath. Hence, I think the MetricRegistry auto-configuration should follow the documented convention and add the appropriate registry.

This is exactly what happens. As you've already noted in your opening description, metrics are being shipped to New Relic. From this we can infer that the New Relic meter registry has been auto-configured as intended.

The problem is not that the meter registry is being switched off but that the registration of some data source-specific meters have been switched off. They are switched off when data source auto-configuration is excluded as it's data source auto-configuration that enables their registration.

As already suggested above, you can avoid the problem by not excluding the data source auto-configuration. Should you wish to exclude the auto-configuration for whatever reason but continue to have DataSource-related metrics exposed, you can import the relevant configuration class:

```java
@Import(DataSourcePoolMetadataProvidersConfiguration.class)
```

Alternatively, you could provide your own `DataSourcePoolMetadataProvider` bean.

Either way, the availability of such a bean will allow `DataSourcePoolMetricsAutoConfiguration` to register meters for the pool's metrics.

**Comment by wilkinsona** [comment]
I've opened https://github.com/spring-projects/spring-boot/issues/13330 for part of what's been discussed in this issue. While the `jdbc.connections` metrics require a `DataSourcePoolMetadataProvider` , there's no reason for the `hikari.connections` metrics to do so.

**Comment by wilkinsona** [comment]
> The Spring Boot Actuator reference document states that each hikari-specific metric is tagged by the name of the pool, which I was not explicitly configuring. I thought this might cause the parsing error, so I configured a pool name for the data source, but this did not solve the error ... although it did change the error message from Error parsing JSON payload at event index 30 to Error parsing JSON payload at event index 8 ... maybe this fixed one parsing error but another remained?

I'm not sure that the change in index is significant as it depends on the batching of the events that are sent. It could have been coincidence that setting the pool name caused the index to change.

I would guess that there's a problem in how Micrometer is creating the JSON that's shipped to New Relic, similar to https://github.com/micrometer-metrics/micrometer/issues/589. Perhaps the fix for that issue is incomplete? It would be useful to know what, if any, custom tag values you are using.

/cc @jkschneider

**Comment by 730alchemy** [comment]
I set a breakpoint in `NewRelicMeterRegistry` `sendEvents` method to examine the request sent to New Relic and the error message received.  It looks like the offending part of the JSON payload sent to NewRelic is the following element:

```
{
	"eventType": "hikaricpConnectionsPending",
	"value": NaN,
	"pool": "local-ds"
}
```

JSON lint raises an error trying to parse the "value" line.  The above element is index 8 in the payload, which matches with the error message.
```
ERROR 448048 --- [pool-1-thread-1] i.m.newrelic.NewRelicMeterRegistry       : failed to send metrics: {"error":"Error parsing JSON payload at event index 8"}
```

RE: custom tags ... I am not using any custom tags

**Comment by wilkinsona** [comment]
Thanks. That looks to me like it should have been addressed by https://github.com/micrometer-metrics/micrometer/commit/3d0f1fd80dc30032c5b31c0de433b1be9d5b1a09. That change is in  Micrometer 1.0.3 so perhaps it hasn't addressed it as you're presumably using 1.0.4.

We've got #13330 for the meter registration side of this issue. I think the remainder needs to be pursued as a Micrometer issue. Unfortunately we can't move an issue from one repository to another. Can you please open a [Micrometer issue](https://github.com/micrometer-metrics/micrometer/issues) for the attempt to ship `NaN`?

**Comment by 730alchemy** [comment]
Thanks Andy for helping me work out the issue of configuring the application to send Hikari metrics. 

Regarding the JSON payload Micrometer sends to New Relic, I dug in further and discovered that part of our code base was pulling in an older version of Micrometer.  When I fixed this to get Micrometer 1.0.4, the issue was resolved.  Our New Relic account is now the proud recipient of Hikari connection pool metrics.
